x-pack/plugins/apm/server/lib/search_strategies/correlations/async_search_service.ts

```diff
@@ -107,11 +107,31 @@ export const asyncSearchServiceProvider = (
         return;
       }
 
+      // finish early if correlation analysis is not required.
+      if (params.analyzeCorrelations === false) {
+        addLogMessage(
+          `Finish service since correlation analysis wasn't requested.`
+        );
+        state.setProgress({
+          loadedHistogramStepsize: 1,
+          loadedOverallHistogram: 1,
+          loadedFieldCanditates: 1,
+          loadedFieldValuePairs: 1,
+          loadedHistograms: 1,
+        });
+        state.setIsRunning(false);
+        return;
+      }
+
       // Create an array of ranges [2, 4, 6, ..., 98]
-      const percents = Array.from(range(2, 100, 2));
+      const percentileAggregationPercents = range(2, 100, 2);
       const {
         percentiles: percentilesRecords,
-      } = await fetchTransactionDurationPercentiles(esClient, params, percents);
+      } = await fetchTransactionDurationPercentiles(
+        esClient,
+        params,
+        percentileAggregationPercents
+      );
```

Comment:
I see `fetchTransactionDurationPercentiles` is called twice. Is it possible to just call it once with the percentiles merged?

Reply:
Another question: in `fetchTransactionDurationPercentiles` why is `percents` optional? Shouldn't it be required?

Reply:
The calls serve different purposes:

- The first time it's called to retrieve the actual value of the 95th percentile. We want to do that call in any case because it ends up as the `95p` marker for the log log chart which can also be retrieved for the trace samples tab without correlation analysis.
- The second time it's called we fetch a different variant of percentiles which after that are used as parameters for correlation analysis which is optional.

Reply:
We plan to revisit the service though, e.g. use msearch/bsearch, nested aggs etc., there are some caveats though regarding to error reporting we need to consider. We have an item "Optimize ES queries (e.g. summarize field/value queries as part of nested aggs or multi search" on the meta issue here: https://github.com/elastic/kibana/issues/106381

Reply:
Regarding `percents` being optional: I don't remember exactly, but maybe the idea was for the helper `getTransactionDurationPercentilesRequest()` to retain the capability of the ES API to use the defaults, maybe we made use of this in an earlier version. We don't use it with defaults at the moment though.

Reply:
As discussed via Slack, pushed a refactor to reduce state handlers here: 1284687